`@taylorize`: add experimental support for `Threads.@threads`

PerezHz opened the pull request:
This PR allows experimental `Threads.@threads` support. Now the user may add `Threads.@threads` before `for` loops in the definition of an ODE when applying the `@taylorize` macro, e.g.:

```julia
@taylorize function f!(dq, q, params, t)
    Threads.@threads for i in 1:10
        dq[i] = q[i]
    end
    nothing
end
```
Also:
```julia
using Base.Threads
@taylorize function f!(dq, q, params, t)
    @threads for i in 1:10 # note there's no `Threads` prefix
        dq[i] = q[i]
    end
    nothing
end
```

I have added some elementary tests which currently only check consistent results on a single thread on travis; right now I'm experimenting with environment variables on travis to actually perform tests on multiple threads.

lbenet commented:
Thanks for this PR!

A naive question: Does this work for ODEs which couple the different degrees of freedom?

PerezHz commented:
> A naive question: Does this work for ODEs which couple the different degrees of freedom?

I've done some locals tests which show that it does; I'll try to add those here to see if they pass on travis

PerezHz commented:
> A naive question: Does this work for ODEs which couple the different degrees of freedom?

@lbenet this is ready for review. I just added a test case with coupled degrees of freedom and things seem to be working alright. I added a couple of `@time`s in the tests just to see how performance is doing on travis, but I'll remove them in the next commits I make here before merging.

PerezHz commented:
Tests are passing on Linux and Windows, but somehow tests are failing on macOS (julia 1.2 and 1.3), I will take a look into this

PerezHz commented:
Tests on julia nightly (macOS) are failing as well...

PerezHz commented:
I was able to fix the error (it was related to an auto-assignment); tests are passing again 👍

PerezHz commented:
Rebased to current master.

lbenet commented:
There seems a conflict with appveyor.yml; probably due to #102. Can you rebase to current master?

lbenet commented on .travis.yml:
```diff
@@ -36,4 +36,5 @@ jobs:
   # push coverage results to Coveralls and Codecov
-  - julia -e 'using Pkg; import TaylorIntegration; joinpath(dirname(pathof(TaylorIntegration)), ".."); Pkg.add("Coverage"); using Coverage; Coveralls.submit(Coveralls.process_folder()); Codecov.submit(Codecov.process_folder())'
+  - julia -e 'using Pkg; Pkg.add("Coverage"); using Coverage; Codecov.submit(process_folder())'
+  - julia -e 'using Pkg; Pkg.add("Coverage"); using Coverage; Coveralls.submit(process_folder())'
```
Don't you need `import TaylorIntegration; joinpath(dirname(pathof(TaylorIntegration)), "..");` in each line, after `using Pkg`?

PerezHz replied:
Instructions in [Coverage.jl](https://github.com/JuliaCI/Coverage.jl#tracking-coverage-with-codecovio) say that it's not necessary. Actually, I was just reading that it may be enough to add `codecov: true` and `coveralls: true` to .travis.yml instead of these two lines.

lbenet replied:
Fair enough; go ahead and make the changes to see :-)

lbenet commented on appveyor.yml:
```diff
@@ -2,7 +2,9 @@ environment:
+  - julia_version: 1.3 # single thread
   - julia_version: 1.3
+    JULIA_NUM_THREADS: 2 # two threads
   - julia_version: nightly
```
(This file needs to be rebase to current master.) 

I think it is advisable to test this functionality in Julia 1.3 and beyond.

PerezHz replied:
Thanks, just rebased to current master

PerezHz replied:
... and added corresponding env variable settings for julia1.4 and nightly

lbenet commented on docs/src/taylorize.md:
```diff
@@ -208,6 +208,8 @@ list some limitations and advices.
 - Use `local` for internal parameters (simple constant values); this improves
   performance. Do not use it if the variable is Taylor expanded.
 
+- `@taylorize` supports multi-threading via `Threads.@threads`.
```
I think we should signal that this is experimental. Do we know if using `@threads` would be thread-safe?

PerezHz replied:
> I think we should signal that this is experimental. 

Ok, I'll add that warning.

>  Do we know if using @threads would be thread-safe?

That is up to the user, since that depends on each particular ODE, so we cannot guarantee thread-safety in advance. Perhaps I should document this as well.

lbenet replied:
Yes, it's important to explicit mention it.

PerezHz replied:
Ok, both warnings are now documented, thanks!

lbenet commented on test/taylorize.jl:
```diff
@@ -5,6 +5,7 @@ using Test
 using LinearAlgebra: norm
 using InteractiveUtils: methodswith
 using Elliptic
+using Base.Threads
```
We may need to add Base.Threads to the `[deps]` section, as well as to `[extras]` and `[targets]` in `Project.toml`.

PerezHz replied:
Thanks, I added Threads in the corresponding sections of Project.toml

lbenet commented on .travis.yml:
```diff
@@ -17,10 +17,10 @@ notifications:
   global:
-    - DOCUMENTER_DEBUG=true
+    - DOCUMENTER_DEBUG=true ## Debug documentation in travis
+    - JULIA_NUM_THREADS=2 ## run Julia on two threads
```
Do you know if there is a *programmatic* way of changing the number of threads? I am thinking in a way to avoid setting an ENV variable...

PerezHz replied:
I'm not aware of a programmatic way to set it programatically... Currently Julia reads the number of threads from the env variable `JULIA_NUM_THREADS`, which must be set *before* Julia starts. It can be set from within Julia, but then the Julia session should be restarted so that the changes take effect.

lbenet replied:
Thanks for the clarification!

PerezHz commented:
Many thanks for taking a look into this, @lbenet!

> I've checked the changes proposed in this PR. It seems that everything is consistent, except for some minor observations. In any case, this PR needs to be rebased to master.

Just rebased to master and pushed a couple of commits which address the main points of your review; please let me know if there is anything that could be improved further.

lbenet commented:
There seems to be a problem with the tests, perhaps related to my proposal of including Threads in Project.toml. @PerezHz, can you remove them and see if things work properly then?

PerezHz commented:
> There seems to be a problem with the tests, perhaps related to my proposal of including Threads in Project.toml. @PerezHz, can you remove them and see if things work properly then?

Sorry for the late reply, and thanks for checking this! Just pushed a commit which reverts the changes in Project.toml; let's see how travis and appveyor react

PerezHz commented:
Tests seem to be passing, except for a codecov error which is not shown by coveralls... should we merge this and possibly tag a new patch/minor version? What do you think @lbenet?

lbenet commented:
Thanks a lot @PerezHz . I'll merge it now.

lbenet commented:
@PerezHz What about tagging a new (patch) version?

PerezHz commented:
Thanks for merging and tagging a new patch version!
